### DISCUSSION TITLE
GraphiQL for self-documenting GraphQL

### DISCUSSION THREAD

**Discussion by kiwicopple** [supabase/supabase#6144] [Feature Requests]
## April 2023 update

This is now supported directly inside the Dashboard: https://supabase.com/blog/supabase-studio-2.0#graphiql

## Previous post

You can use GraphiQL with `pg_graphql` for a self-documenting UI, similar to our REST docs. This is already provided by GraphiQL, however our endpoints require an `apikey` header. One way to achieve this is to create a local file `graphiql.html` with the following contents:

```html
<html>
  <head>
    <title>GraphiQL</title>
    <link href="https://unpkg.com/graphiql/graphiql.css" rel="stylesheet" />
  </head>
  <body style="margin: 0;">
    <div id="graphiql" style="height: 100vh;"></div>

    <script
      crossorigin
      src="https://unpkg.com/react/umd/react.production.min.js"
    ></script>
    <script
      crossorigin
      src="https://unpkg.com/react-dom/umd/react-dom.production.min.js"
    ></script>
    <script
      crossorigin
      src="https://unpkg.com/graphiql/graphiql.js"
    ></script>

    <script>
      const fetcher = GraphiQL.createFetcher({
              url: 'https://<project_ref>.supabase.co/graphql/v1',
             headers: {
                 "apiKey": "<apiKey>"
             }
      });
      ReactDOM.render(
        React.createElement(GraphiQL, { fetcher: fetcher }),
        document.getElementById('graphiql'),
      );
    </script>
  </body>
</html>

```

Replace `<project_ref>` and `<apiKey>` with your own details, open it in a browser, and you have local-hosted GraphiQL docs 👍

**Comment by jaytxng** [comment]
This is awesome to get going! 

I notice that it can sometimes take awhile for those packages between refreshes. I made a [supabase_graphiql](https://www.npmjs.com/package/supabase_graphiql) tool for our teams to avoid those subsequent network calls.

**Reply by acao** [reply]
hey there @jaytxng - graphiql co-maintainer and fellow supabase user here - can you [report this](https://github.com/graphql/graphiql/issues/new?assignees=&labels=bug%2Cneeds+triage%2Cgraphiql&template=graphiql-bug.yml&title=%5Bgraphiql%5D+%3Ctitle%3E) and the specific UX issues? I would love to use this bug to set a target for bundle size that makes TTI more reasonable on refresh when network cacheing is disabled.

I will take the blame for this - in late 1.x and early 2.x, we added `@graphiql/react` but decided to use vite lib mode, and this leads to the dependencies being bundled almost in duplicate when graphiql package is built!  Personally I don't think module libraries should bundle dependencies in most cases, that's what CDN bundles are for! Overhauling the build toolchain will take some time, perhaps a few more months as we have time to get to it.  We've found tsup to be an excellent replacement for vite for module libraries, and instead the graphiql.min.js CDN bundle will be the only lib bundled with deps by vite lib mode, as it should be. (yes you can mark all runtime deps as external in theory, but we couldn't get this working with vite, including the plugin made for this :()

**Reply by acao** [reply]
@jaytxng is there a reason you are doing full page refreshes? there is a schema refresh button and shortcut

**Reply by jaytxng** [reply]
Sorry for the late reply. I personally was fine with the schema refresh button, but other teammates have tab suspenders/other extensions that forces a full page refresh when coming back to the tab.

As an FYI, we've switched over to use the in-supabase graphiql tool and don't have this issue anymore.
